src/backend/distributed/planner/recursive_planning.c

```diff
@@ -0,0 +1,606 @@
+								 PlannerRestrictionContext *plannerRestrictionContext,
+								 uint64 planId, List **subPlanList)
+{
+	DeferredErrorMessage *error = NULL;
+	RecursivePlanningContext context;
+
+	if (SubqueryPushdown)
+	{
+		/*
+		 * When the subquery_pushdown flag is enabled we make some hacks
+		 * to push down subqueries with LIMIT. Recursive planning would
+		 * valiantly do the right thing and try to recursively plan the
+		 * inner subqueries, but we don't really want it to because those
+		 * subqueries might not be supported and would be much slower.
+		 *
+		 * Instead, we skip recursive planning altogether when
+		 * subquery_pushdown is enabled.
+		 */
+		return false;
+	}
+
+	context.level = 0;
+	context.planId = planId;
+	context.subPlanList = NIL;
+	context.plannerRestrictionContext = plannerRestrictionContext;
+
+	error = RecursivelyPlanCTEs(query, &context);
+	if (error != NULL)
+	{
+		return error;
+	}
+
+	/* XXX: plan subqueries */
+
+	*subPlanList = context.subPlanList;
+
+	return NULL;
+}
+
+
+/*
+ * RecursivelyPlanCTEs plans all CTEs in the query by recursively calling the planner
+ * The resulting plan is added to planningContext->subPlanList and CTE references
+ * are replaced by subqueries that call read_intermediate_result, which reads the
+ * intermediate result of the CTE after it is executed.
+ *
+ * Recursive and modifying CTEs are not yet supported and return an error.
+ */
+static DeferredErrorMessage *
+RecursivelyPlanCTEs(Query *query, RecursivePlanningContext *planningContext)
+{
+	ListCell *cteCell = NULL;
+	CteReferenceWalkerContext context = { -1, NIL };
+
+	if (query->hasModifyingCTE)
+	{
+		/* we could easily support these, but it's a little scary */
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "data-modifying statements are not supported in "
+							 "the WITH clauses of distributed queries",
+							 NULL, NULL);
+	}
+
+	if (query->hasRecursive)
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "recursive CTEs are not supported in distributed "
+							 "queries",
+							 NULL, NULL);
+	}
+
+	/* get all RTE_CTEs that point to CTEs from cteList */
+	CteReferenceListWalker((Node *) query, &context);
+
+	foreach(cteCell, query->cteList)
+	{
+		CommonTableExpr *cte = (CommonTableExpr *) lfirst(cteCell);
+		char *cteName = cte->ctename;
+		Query *subquery = (Query *) cte->ctequery;
+		Query *subPlanQuery = copyObject(subquery);
+		uint64 planId = planningContext->planId;
+		uint32 subPlanId = list_length(planningContext->subPlanList);
+		Query *resultQuery = NULL;
+		DistributedSubPlan *subPlan = NULL;
+		ListCell *rteCell = NULL;
+
+		if (ContainsReferencesToOuterQuery(subquery))
+		{
+			return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+								 "CTEs that refer to other subqueries are not "
+								 "supported in multi-shard queries",
+								 NULL, NULL);
+		}
+
+		/* we want to be able to handle queries with only intermediate results */
```

Comment:
Removed the check from here. I added it because there were a lot of subqueries in the regression tests that started using recursive planning, but then give very confusing error messages because they have router executor disabled.